src/main/java/com/ibm/crypto/plus/provider/XDHPrivateKeyImpl.java

```diff
@@ -313,25 +313,60 @@ private byte[] processEncodedPrivateKey(byte[] encoded) throws IOException {
         DerInputStream derStream = new DerInputStream(keyBytes);
         try {
-            // XDH private key in SunEC new Java 17 design requires [octet-string[octer-string[key-bytes]]] format,
-            // otherwise, it causes interop issue. JCK issue 569
-            this.key = derStream.getOctetString(); // Try J17 format [octet-string[octer-string[key-bytes]]]
-        } catch (IOException e) {
-            this.key = keyBytes; // Try J11 format [octer-string[key-bytes]]
+            // XDH private key in SunEC new Java 17 design requires [octet-string[octet-string[key-bytes]]] format,
+            // otherwise, it causes interop issue.
+            if (isCorrectlyFormedOctetString(keyBytes)) {
+                this.key = derStream.getOctetString(); // We know we are working with the format [octet-string[octet-string[key-bytes]]]
+            } else {
+                throw new IOException("Octet is malformed, use entire key byte array.");
+            }
+        } catch (Exception e) {
+            //e.printStackTrace();
+            this.key = keyBytes; // Try J11 format [octet-string[key-bytes]]
         }
-        DerOutputStream encodedKey = new DerOutputStream();
-        if (CurveUtil.isFFDHE(this.curve)) {
-            BigInteger octetStringAsBigInt = new BigInteger(this.key);
-            encodedKey.putInteger(octetStringAsBigInt); // Put in another octet string
-        } else {
-            encodedKey.putOctetString(this.key); // Put in another octet string
+        try (DerOutputStream encodedKey = new DerOutputStream()) {
+            if (CurveUtil.isFFDHE(this.curve)) {
+                BigInteger octetStringAsBigInt = new BigInteger(this.key);
+                encodedKey.putInteger(octetStringAsBigInt); // Put in another octet string
+            } else {
+                encodedKey.putOctetString(this.key); // Put in another octet string
+            }
+            outStream.putOctetString(encodedKey.toByteArray());
         }
-        outStream.putOctetString(encodedKey.toByteArray());
 
-        DerOutputStream asn1Key = new DerOutputStream();
-        asn1Key.write(DerValue.tag_Sequence, outStream);
+        try (DerOutputStream asn1Key = new DerOutputStream()) {
+            asn1Key.write(DerValue.tag_Sequence, outStream);
+            return asn1Key.toByteArray();
+        }
+    }
+
+
+    /**
+     * Determines if a given array is a properly formed DER octet string.
+     * 
+     * @param keyBytes The byte array to check for a complete octet.
+     * @return Returns true if the first byte of the array indicates an octet ( 0x04 ) and the
+     * value continues to the end of the array indicating a complete octet filling the 
+     * given byte array. Returns false otherwise.
+     * @throws IOException Throws an IOException when a failure occurs decoding the DER encoded bytes.
+     */
+    private boolean isCorrectlyFormedOctetString(byte[] keyBytes) throws Exception {
```

Comment:
I think we should specify the exact type of exception expected.

Reply:
Agreed will update to use IOException.